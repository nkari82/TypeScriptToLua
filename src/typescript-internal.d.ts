export {};

declare module "typescript" {
    function createDiagnosticReporter(system: System, pretty?: boolean): DiagnosticReporter;
    function createWatchStatusReporter(system: System, pretty?: boolean): WatchStatusReporter;

    interface System {
        setBlocking?(): void;
    }

    interface Statement {
        jsDoc?: JSDoc[];
    }

    interface Program {
        getCommonSourceDirectory(): string;
    }

    interface CompilerOptions {
        configFile?: TsConfigSourceFile;
        configFilePath?: string;
    }

    interface TypeChecker {
        getElementTypeOfArrayType(type: Type): Type | undefined;
        getContextualTypeForObjectLiteralElement(element: ObjectLiteralElementLike): Type | undefined;

        isTupleType(type: Type): boolean;
        isArrayType(type: Type): boolean;
    }

    interface Symbol {
        // false positive
        // eslint-disable-next-line @typescript-eslint/ban-types
        parent?: Symbol;
    }

    function transformJsx(context: TransformationContext): (x: SourceFile) => SourceFile;

    export type OuterExpression =
        | ParenthesizedExpression
        | TypeAssertion
        | AsExpression
        | NonNullExpression
        | PartiallyEmittedExpression;

    function skipOuterExpressions(node: Expression, kinds?: OuterExpressionKinds): Expression;
    export function isOuterExpression(node: Node, kinds?: OuterExpressionKinds): node is OuterExpression;
}
