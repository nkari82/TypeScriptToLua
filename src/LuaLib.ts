import * as path from "path";
import { EmitHost } from "./transpilation";
import * as lua from "./LuaAST";

export enum LuaLibFeature {
    ArrayConcat = "ArrayConcat",
    ArrayEntries = "ArrayEntries",
    ArrayEvery = "ArrayEvery",
    ArrayFilter = "ArrayFilter",
    ArrayForEach = "ArrayForEach",
    ArrayFind = "ArrayFind",
    ArrayFindIndex = "ArrayFindIndex",
    ArrayFrom = "ArrayFrom",
    ArrayIncludes = "ArrayIncludes",
    ArrayIndexOf = "ArrayIndexOf",
    ArrayIsArray = "ArrayIsArray",
    ArrayJoin = "ArrayJoin",
    ArrayMap = "ArrayMap",
    ArrayPush = "ArrayPush",
    ArrayPushArray = "ArrayPushArray",
    ArrayReduce = "ArrayReduce",
    ArrayReduceRight = "ArrayReduceRight",
    ArrayReverse = "ArrayReverse",
    ArrayUnshift = "ArrayUnshift",
    ArraySort = "ArraySort",
    ArraySlice = "ArraySlice",
    ArraySome = "ArraySome",
    ArraySplice = "ArraySplice",
    ArrayToObject = "ArrayToObject",
    ArrayFlat = "ArrayFlat",
    ArrayFlatMap = "ArrayFlatMap",
    ArraySetLength = "ArraySetLength",
    Await = "Await",
    Class = "Class",
    ClassExtends = "ClassExtends",
    CloneDescriptor = "CloneDescriptor",
    Decorate = "Decorate",
    DecorateParam = "DecorateParam",
    Delete = "Delete",
    DelegatedYield = "DelegatedYield",
    Error = "Error",
    FunctionBind = "FunctionBind",
    Generator = "Generator",
    InstanceOf = "InstanceOf",
    InstanceOfObject = "InstanceOfObject",
    Iterator = "Iterator",
    LuaIteratorSpread = "LuaIteratorSpread",
    Map = "Map",
    MathAtan2 = "MathAtan2",
    MathSign = "MathSign",
    New = "New",
    Number = "Number",
    NumberIsFinite = "NumberIsFinite",
    NumberIsNaN = "NumberIsNaN",
    NumberToString = "NumberToString",
    ObjectAssign = "ObjectAssign",
    ObjectDefineProperty = "ObjectDefineProperty",
    ObjectEntries = "ObjectEntries",
    ObjectFromEntries = "ObjectFromEntries",
    ObjectGetOwnPropertyDescriptor = "ObjectGetOwnPropertyDescriptor",
    ObjectGetOwnPropertyDescriptors = "ObjectGetOwnPropertyDescriptors",
    ObjectKeys = "ObjectKeys",
    ObjectRest = "ObjectRest",
    ObjectValues = "ObjectValues",
    ParseFloat = "ParseFloat",
    ParseInt = "ParseInt",
    Promise = "Promise",
    PromiseAll = "PromiseAll",
    PromiseAllSettled = "PromiseAllSettled",
    PromiseAny = "PromiseAny",
    PromiseRace = "PromiseRace",
    Set = "Set",
    SetDescriptor = "SetDescriptor",
    SparseArrayNew = "SparseArrayNew",
    SparseArrayPush = "SparseArrayPush",
    SparseArraySpread = "SparseArraySpread",
    WeakMap = "WeakMap",
    WeakSet = "WeakSet",
    SourceMapTraceBack = "SourceMapTraceBack",
    Spread = "Spread",
    StringAccess = "StringAccess",
    StringCharAt = "StringCharAt",
    StringCharCodeAt = "StringCharCodeAt",
    StringEndsWith = "StringEndsWith",
    StringIncludes = "StringIncludes",
    StringPadEnd = "StringPadEnd",
    StringPadStart = "StringPadStart",
    StringReplace = "StringReplace",
    StringReplaceAll = "StringReplaceAll",
    StringSlice = "StringSlice",
    StringSplit = "StringSplit",
    StringStartsWith = "StringStartsWith",
    StringSubstr = "StringSubstr",
    StringSubstring = "StringSubstring",
    StringTrim = "StringTrim",
    StringTrimEnd = "StringTrimEnd",
    StringTrimStart = "StringTrimStart",
    Symbol = "Symbol",
    SymbolRegistry = "SymbolRegistry",
    TypeOf = "TypeOf",
    Unpack = "Unpack",
}

export interface LuaLibFeatureInfo {
    dependencies?: LuaLibFeature[];
    exports: string[];
}
export type LuaLibModulesInfo = Record<LuaLibFeature, LuaLibFeatureInfo>;

export const luaLibModulesInfoFileName = "lualib_module_info.json";
let luaLibModulesInfo: LuaLibModulesInfo | undefined;
export function getLuaLibModulesInfo(emitHost: EmitHost): LuaLibModulesInfo {
    if (luaLibModulesInfo === undefined) {
        const lualibPath = path.resolve(__dirname, `../dist/lualib/${luaLibModulesInfoFileName}`);
        const result = emitHost.readFile(lualibPath);
        if (result !== undefined) {
            luaLibModulesInfo = JSON.parse(result) as LuaLibModulesInfo;
        } else {
            throw new Error(`Could not load lualib dependencies from '${lualibPath}'`);
        }
    }
    return luaLibModulesInfo;
}

export function readLuaLibFeature(feature: LuaLibFeature, emitHost: EmitHost): string {
    const featurePath = path.resolve(__dirname, `../dist/lualib/${feature}.lua`);
    const luaLibFeature = emitHost.readFile(featurePath);
    if (luaLibFeature === undefined) {
        throw new Error(`Could not load lualib feature from '${featurePath}'`);
    }
    return luaLibFeature;
}

export function resolveRecursiveLualibFeatures(
    features: Iterable<LuaLibFeature>,
    emitHost: EmitHost,
    luaLibModulesInfo: LuaLibModulesInfo = getLuaLibModulesInfo(emitHost)
): LuaLibFeature[] {
    const loadedFeatures = new Set<LuaLibFeature>();
    const result: LuaLibFeature[] = [];

    function load(feature: LuaLibFeature): void {
        if (loadedFeatures.has(feature)) return;
        loadedFeatures.add(feature);

        const dependencies = luaLibModulesInfo[feature]?.dependencies;
        if (dependencies) {
            dependencies.forEach(load);
        }

        result.push(feature);
    }

    for (const feature of features) {
        load(feature);
    }

    return result;
}

export function loadInlineLualibFeatures(features: Iterable<LuaLibFeature>, emitHost: EmitHost): string {
    let result = "";

    for (const feature of resolveRecursiveLualibFeatures(features, emitHost)) {
        const luaLibFeature = readLuaLibFeature(feature, emitHost);
        result += luaLibFeature + "\n";
    }

    return result;
}

export function loadImportedLualibFeatures(features: Iterable<LuaLibFeature>, emitHost: EmitHost): lua.Statement[] {
    const luaLibModuleInfo = getLuaLibModulesInfo(emitHost);

    const imports = Array.from(features).flatMap(feature => luaLibModuleInfo[feature].exports);

    const requireCall = lua.createCallExpression(lua.createIdentifier("require"), [
        lua.createStringLiteral("lualib_bundle"),
    ]);
    if (imports.length === 0) {
        return [];
    }

    const luaLibId = lua.createIdentifier("____lualib");
    const importStatement = lua.createVariableDeclarationStatement(luaLibId, requireCall);
    const statements: lua.Statement[] = [importStatement];
    // local <export> = ____luaLib.<export>
    for (const item of imports) {
        statements.push(
            lua.createVariableDeclarationStatement(
                lua.createIdentifier(item),
                lua.createTableIndexExpression(luaLibId, lua.createStringLiteral(item))
            )
        );
    }
    return statements;
}

let luaLibBundleContent: string;
export function getLuaLibBundle(emitHost: EmitHost): string {
    if (luaLibBundleContent === undefined) {
        const lualibPath = path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua");
        const result = emitHost.readFile(lualibPath);
        if (result !== undefined) {
            luaLibBundleContent = result;
        } else {
            throw new Error(`Could not load lualib bundle from '${lualibPath}'`);
        }
    }

    return luaLibBundleContent;
}
