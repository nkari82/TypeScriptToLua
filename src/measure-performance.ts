import { performance } from "perf_hooks";

// We use our own performance hooks implementation for easier use, but also call node's performance hooks, so it shows up in the profiler.

let enabled = false;
const marks = new Map<string, number>();
const durations = new Map<string, number>();

function timestamp() {
    return performance.now();
}

/**
 * Marks a performance event, with the given markName.
 */
function mark(markName: string) {
    if (enabled) {
        marks.set(markName, timestamp());
        performance.mark(markName);
    }
}

/**
 * Adds a performance measurement with the specified name.
 *
 * @param measureName The name of the performance measurement.
 * @param startMarkName The name of the starting mark
 * @param endMarkName The name of the ending mark
 */
function measure(measureName: string, startMarkName: string, endMarkName: string) {
    if (enabled) {
        const end = marks.get(endMarkName) ?? timestamp();
        const start = marks.get(startMarkName) ?? performance.timeOrigin;
        const previousDuration = durations.get(measureName) ?? 0;
        durations.set(measureName, previousDuration + (end - start));
        performance.measure(measureName, startMarkName, endMarkName);
    }
}

/**
 * Starts a performance measurement section.
 * @param name name of the measurement
 */
export function startSection(name: string) {
    mark("start " + name);
}

/**
 * Ends a performance measurement section.
 * @param name name of the measurement
 */
export function endSection(name: string) {
    mark("end " + name);
    measure(name, "start " + name, "end " + name);
}

export function isMeasurementEnabled() {
    return enabled;
}

export function enableMeasurement() {
    if (!enabled) {
        enabled = true;
    }
}

export function disableMeasurement() {
    if (enabled) {
        enabled = false;
        marks.clear();
        durations.clear();
    }
}

export function forEachMeasure(callback: (measureName: string, duration: number) => void) {
    durations.forEach((duration, measureName) => callback(measureName, duration));
}

export function getTotalDuration() {
    let total = 0;
    forEachMeasure((_, duration) => (total += duration));
    return total;
}
