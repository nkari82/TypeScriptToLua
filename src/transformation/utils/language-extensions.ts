import * as ts from "typescript";
import { TransformationContext } from "../context";

export enum ExtensionKind {
    MultiFunction = "MultiFunction",
    RangeFunction = "RangeFunction",
    VarargConstant = "VarargConstant",
    AdditionOperatorType = "Addition",
    AdditionOperatorMethodType = "AdditionMethod",
    SubtractionOperatorType = "Subtraction",
    SubtractionOperatorMethodType = "SubtractionMethod",
    MultiplicationOperatorType = "Multiplication",
    MultiplicationOperatorMethodType = "MultiplicationMethod",
    DivisionOperatorType = "Division",
    DivisionOperatorMethodType = "DivisionMethod",
    ModuloOperatorType = "Modulo",
    ModuloOperatorMethodType = "ModuloMethod",
    PowerOperatorType = "Power",
    PowerOperatorMethodType = "PowerMethod",
    FloorDivisionOperatorType = "FloorDivision",
    FloorDivisionOperatorMethodType = "FloorDivisionMethod",
    BitwiseAndOperatorType = "BitwiseAnd",
    BitwiseAndOperatorMethodType = "BitwiseAndMethod",
    BitwiseOrOperatorType = "BitwiseOr",
    BitwiseOrOperatorMethodType = "BitwiseOrMethod",
    BitwiseExclusiveOrOperatorType = "BitwiseExclusiveOr",
    BitwiseExclusiveOrOperatorMethodType = "BitwiseExclusiveOrMethod",
    BitwiseLeftShiftOperatorType = "BitwiseLeftShift",
    BitwiseLeftShiftOperatorMethodType = "BitwiseLeftShiftMethod",
    BitwiseRightShiftOperatorType = "BitwiseRightShift",
    BitwiseRightShiftOperatorMethodType = "BitwiseRightShiftMethod",
    ConcatOperatorType = "Concat",
    ConcatOperatorMethodType = "ConcatMethod",
    LessThanOperatorType = "LessThan",
    LessThanOperatorMethodType = "LessThanMethod",
    GreaterThanOperatorType = "GreaterThan",
    GreaterThanOperatorMethodType = "GreaterThanMethod",
    NegationOperatorType = "Negation",
    NegationOperatorMethodType = "NegationMethod",
    BitwiseNotOperatorType = "BitwiseNot",
    BitwiseNotOperatorMethodType = "BitwiseNotMethod",
    LengthOperatorType = "Length",
    LengthOperatorMethodType = "LengthMethod",
    TableNewType = "TableNew",
    TableDeleteType = "TableDelete",
    TableDeleteMethodType = "TableDeleteMethod",
    TableGetType = "TableGet",
    TableGetMethodType = "TableGetMethod",
    TableHasType = "TableHas",
    TableHasMethodType = "TableHasMethod",
    TableSetType = "TableSet",
    TableSetMethodType = "TableSetMethod",
    TableAddKeyType = "TableAddKey",
    TableAddKeyMethodType = "TableAddKeyMethod",
}
const extensionValues: Set<string> = new Set(Object.values(ExtensionKind));

export function getExtensionKindForType(context: TransformationContext, type: ts.Type): ExtensionKind | undefined {
    const value = getPropertyValue(context, type, "__tstlExtension");
    if (value && extensionValues.has(value)) {
        return value as ExtensionKind;
    }
}

const excludedTypeFlags: ts.TypeFlags =
    ((1 << 18) - 1) | // All flags from Any...Never
    ts.TypeFlags.Index |
    ts.TypeFlags.NonPrimitive;

function getPropertyValue(context: TransformationContext, type: ts.Type, propertyName: string): string | undefined {
    if (type.flags & excludedTypeFlags) return;
    const property = type.getProperty(propertyName);
    if (!property) return undefined;
    const propertyType = context.checker.getTypeOfSymbolAtLocation(property, context.sourceFile);
    if (propertyType.isStringLiteral()) return propertyType.value;
}

export function getExtensionKindForNode(context: TransformationContext, node: ts.Node): ExtensionKind | undefined {
    const originalNode = ts.getOriginalNode(node);
    let type = context.checker.getTypeAtLocation(originalNode);
    if (ts.isOptionalChain(originalNode)) {
        type = context.checker.getNonNullableType(type);
    }
    return getExtensionKindForType(context, type);
}

export function getExtensionKindForSymbol(
    context: TransformationContext,
    symbol: ts.Symbol
): ExtensionKind | undefined {
    const type = context.checker.getTypeOfSymbolAtLocation(symbol, context.sourceFile);
    return getExtensionKindForType(context, type);
}

export enum IterableExtensionKind {
    Iterable = "Iterable",
    Pairs = "Pairs",
    PairsKey = "PairsKey",
}

export function getIterableExtensionTypeForType(
    context: TransformationContext,
    type: ts.Type
): IterableExtensionKind | undefined {
    const value = getPropertyValue(context, type, "__tstlIterable");
    if (value && value in IterableExtensionKind) {
        return value as IterableExtensionKind;
    }
}

export function getIterableExtensionKindForNode(
    context: TransformationContext,
    node: ts.Node
): IterableExtensionKind | undefined {
    const type = context.checker.getTypeAtLocation(node);
    return getIterableExtensionTypeForType(context, type);
}
