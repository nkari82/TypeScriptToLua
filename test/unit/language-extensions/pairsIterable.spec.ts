import * as util from "../../util";
import { invalidPairsIterableWithoutDestructuring } from "../../../src/transformation/utils/diagnostics";
import { LuaTarget } from "../../../src";

const testIterable = `
const testIterable = {a1: "a2", b1: "b2", c1: "c2"} as unknown as LuaPairsIterable<string, string>;
`;

const testResults = {
    a1: "a2",
    b1: "b2",
    c1: "c2",
};

test("pairs iterable", () => {
    util.testFunction`
        ${testIterable}
        const results: Record<string, string> = {};
        for (const [k, v] of testIterable) {
            results[k] = v;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual(testResults);
});

test("pairs iterable with external control variable", () => {
    util.testFunction`
        ${testIterable}
        const results: Record<string, string> = {};
        let k: string, v: string;
        for ([k, v] of testIterable) {
            results[k] = v;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual(testResults);
});

test("pairs iterable function forward", () => {
    util.testFunction`
        ${testIterable}
        function forward() { return testIterable; }
        const results: Record<string, string> = {};
        for (const [k, v] of forward()) {
            results[k] = v;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual(testResults);
});

test("pairs iterable function indirect forward", () => {
    util.testFunction`
        ${testIterable}
        function forward() { const iter = testIterable; return iter; }
        const results: Record<string, string> = {};
        for (const [k, v] of forward()) {
            results[k] = v;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual(testResults);
});

test("pairs iterable arrow function forward", () => {
    util.testFunction`
        ${testIterable}
        const forward = () => testIterable;
        const results: Record<string, string> = {};
        for (const [k, v] of forward()) {
            results[k] = v;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual(testResults);
});

test("pairs iterable with __pairs metamethod", () => {
    util.testFunction`
        class PairsTest  {
            __pairs() {
                const kvp = [ ["a1", "a2"], ["b1", "b2"], ["c1", "c2"] ];
                let i = 0;
                return () => {
                    if (i < kvp.length) {
                        const [k, v] = kvp[i++];
                        return $multi(k, v);
                    }
                };
            }
        }
        const tester = new PairsTest() as PairsTest & LuaPairsIterable<string, string>;
        const results: Record<string, string> = {};
        for (const [k, v] of tester) {
            results[k] = v;
        }
        return results;
    `
        .withLanguageExtensions()
        .setOptions({ luaTarget: LuaTarget.Lua53 })
        .expectToEqual(testResults);
});

test.each(["for (const s of testIterable) {}", "let s; for (s of testIterable) {}"])(
    "invalid LuaPairsIterable without destructuring (%p)",
    statement => {
        util.testFunction`
        ${testIterable}
        ${statement}
    `
            .withLanguageExtensions()
            .expectDiagnosticsToMatchSnapshot([invalidPairsIterableWithoutDestructuring.code]);
    }
);

const testKeyIterable = `
const testKeyIterable = {a1: true, b1: true, c1: true} as unknown as LuaPairsKeyIterable<string>;
`;

test("pairs key iterable", () => {
    util.testFunction`
        ${testKeyIterable}
        const results: Record<string, boolean> = {};
        for (const k of testKeyIterable) {
            results[k] = true;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual({ a1: true, b1: true, c1: true });
});

test("pairs key iterable with external control variable", () => {
    util.testFunction`
        ${testKeyIterable}
        const results: Record<string, boolean> = {};
        let k: string;
        for (k of testKeyIterable) {
            results[k] = true;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual({ a1: true, b1: true, c1: true });
});

test("pairs key iterable function forward", () => {
    util.testFunction`
        ${testKeyIterable}
        function forward() { return testKeyIterable; }
        const results: Record<string, boolean> = {};
        for (const k of forward()) {
            results[k] = true;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual({ a1: true, b1: true, c1: true });
});

test("pairs key iterable function indirect forward", () => {
    util.testFunction`
        ${testKeyIterable}
        function forward() { const iter = testKeyIterable; return iter; }
        const results: Record<string, boolean> = {};
        for (const k of forward()) {
            results[k] = true;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual({ a1: true, b1: true, c1: true });
});

test("pairs key iterable arrow function forward", () => {
    util.testFunction`
        ${testKeyIterable}
        const forward = () => testKeyIterable;
        const results: Record<string, boolean> = {};
        for (const k of forward()) {
            results[k] = true;
        }
        return results;
    `
        .withLanguageExtensions()
        .expectToEqual({ a1: true, b1: true, c1: true });
});

test("pairs key iterable with __pairs metamethod", () => {
    util.testFunction`
        class PairsTest  {
            __pairs() {
                const kvp = [ ["a1", true], ["b1", true], ["c1", true] ];
                let i = 0;
                return () => {
                    if (i < kvp.length) {
                        const [k, v] = kvp[i++];
                        return $multi(k, v);
                    }
                };
            }
        }
        const tester = new PairsTest() as PairsTest & LuaPairsKeyIterable<string>;
        const results: Record<string, boolean> = {};
        for (const k of tester) {
            results[k] = true;
        }
        return results;
    `
        .withLanguageExtensions()
        .setOptions({ luaTarget: LuaTarget.Lua53 })
        .expectToEqual({ a1: true, b1: true, c1: true });
});
