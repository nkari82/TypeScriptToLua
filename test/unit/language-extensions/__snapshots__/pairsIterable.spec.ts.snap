// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid LuaPairsIterable without destructuring ("for (const s of testIterable) {}"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local testIterable = {a1 = "a2", b1 = "b2", c1 = "c2"}
    for ____ in pairs(testIterable) do
    end
end
return ____exports"
`;

exports[`invalid LuaPairsIterable without destructuring ("for (const s of testIterable) {}"): diagnostics 1`] = `"main.ts(5,20): error TSTL: LuaPairsIterable type must be destructured in a for...of statement."`;

exports[`invalid LuaPairsIterable without destructuring ("let s; for (s of testIterable) {}"): code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local testIterable = {a1 = "a2", b1 = "b2", c1 = "c2"}
    local s
    for ____ in pairs(testIterable) do
    end
end
return ____exports"
`;

exports[`invalid LuaPairsIterable without destructuring ("let s; for (s of testIterable) {}"): diagnostics 1`] = `"main.ts(5,21): error TSTL: LuaPairsIterable type must be destructured in a for...of statement."`;
